cdap-app-fabric/src/main/java/co/cask/cdap/lineage/field/FieldLineageInfo.java

```diff
@@ -138,10 +161,120 @@ public long getChecksum() {
+
   private long computeChecksum() {
     return fingerprint64(canonicalize().getBytes(Charsets.UTF_8));
   }
 
+  private Map<EndPoint, Set<String>> computeDestinationFields() {
+    Map<EndPoint, Set<String>> destinationFields = new HashMap<>();
+    for (Operation operation : this.operations) {
+      if (operation.getType() != OperationType.WRITE) {
+        continue;
+      }
+
+      WriteOperation write = (WriteOperation) operation;
+      Set<String> endPointFields = destinationFields.computeIfAbsent(write.getDestination(), k -> new HashSet<>());
+      for (InputField field : write.getInputs()) {
+        endPointFields.add(field.getName());
+      }
+    }
+    return destinationFields;
+  }
+
+  private Map<EndPointField, Set<EndPointField>> computeIncomingSummary() {
+    Map<EndPointField, Set<EndPointField>> result = new HashMap<>();
+    for (WriteOperation write : writeOperations) {
+      List<InputField> inputs = write.getInputs();
+      for (InputField input : inputs) {
+        Set<String> visitedOperationNames = new HashSet<>();
+        computeIncomingSummaryHelper(new EndPointField(write.getDestination(), input.getName()),
+                                     operationsMap.get(input.getOrigin()), write, visitedOperationNames, result);
+      }
+    }
+    return result;
+  }
+
+  private void computeIncomingSummaryHelper(EndPointField destination, Operation currentOperation,
+                                            Operation previousOperation, Set<String> visitedOperationNames,
+                                            Map<EndPointField, Set<EndPointField>> result) {
+    if (!visitedOperationNames.add(currentOperation.getName())) {
+      // cycles detected
+      // TODO: CDAP-13548
+    }
+
+    if (currentOperation.getType() == OperationType.READ) {
+      // if current operation is of type READ, previous operation must be of type TRANSFORM or WRITE
+      // get only the input fields from the previous operations for which the origin is current READ operation
+      Set<InputField> inputFields = new HashSet<>();
+      if (OperationType.WRITE == previousOperation.getType()) {
+        WriteOperation previousWrite = (WriteOperation) previousOperation;
+        inputFields = new HashSet<>(previousWrite.getInputs());
+      } else if (OperationType.TRANSFORM == previousOperation.getType()) {
+        TransformOperation previousTransform = (TransformOperation) previousOperation;
+        inputFields = new HashSet<>(previousTransform.getInputs());
+      }
+
+      Set<EndPointField> sourceEndPointFields = result.computeIfAbsent(destination, k -> new HashSet<>());
+
+      ReadOperation read = (ReadOperation) currentOperation;
+      EndPoint source = read.getSource();
+      for (InputField inputField : inputFields) {
+        if (inputField.getOrigin().equals(currentOperation.getName())) {
+          sourceEndPointFields.add(new EndPointField(source, inputField.getName()));
+        }
+      }
+
+      return;
```

Comment:
redundant return?

Reply:
`return` is required here since we assume the operation is of type Transform on line 248.

Reply:
If you remove the return here, it will still return a few lines down from here. But perhaps it is clearer and more resilient to future change to leave it here.